[Clang] Add support for fp when using min_fetch/max_fetch atomics

Amichaxx opened the pull request:
Previously when using min_fetch/max_fetch atomics with floating point types, LLVM would emit a crash. This patch updates the EmitPostAtomicMinMax function in CGAtomic.cpp to take floating point types. Included is a clang CodeGen test atomic-ops-float-check-minmax.c and Sema test atomic-ops-fp-minmax.c.

Amichaxx committed:
[Clang] Add support for fp when using min_fetch/max_fetch atomics

Previously when using min_fetch/max_fetch atomics with floating point types, LLVM would emit a crash. This patch updates the EmitPostAtomicMinMax function in CGAtomic.cpp to take floating point types. Included is a clang CodeGen test atomic-ops-float-check-minmax.c

CarolineConcatto reviewed (commented):
I saw in this patch:

```
commit 5cf58768cb3ba31ee37facaf23f7a74f78781590
Author: Tim Northover <t.p.northover@gmail.com>
Date:   Thu Nov 21 10:31:30 2019 +0000
```

that they also added tests Sema tests:
clang/test/Sema/atomic-ops.c
Can we do the same?

I am not sure why it was not added support for floating point at the in that patch, maybe it would be nice to add as well Tim Northover as a reviewer.

CarolineConcatto commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -507,20 +507,35 @@ static llvm::Value *EmitPostAtomicMinMax(CGBuilderTy &Builder,
   case AtomicExpr::AO__scoped_atomic_max_fetch:
-    Pred = IsSigned ? llvm::CmpInst::ICMP_SGT : llvm::CmpInst::ICMP_UGT;
+    Pred = IsFP
+               ? llvm::CmpInst::FCMP_OGT
```
Hi @efriedma-quic ,
Thank you for the review: 

Are you talking about this:
> fmax: *ptr = maxnum(*ptr, val) (match the llvm.maxnum.* intrinsic)
> fmin: *ptr = minnum(*ptr, val) (match the llvm.minnum.* intrinsic)
> fmaximum: *ptr = maximum(*ptr, val) (match the llvm.maximum.* intrinsic)
> fminimum: *ptr = minimum(*ptr, val) (match the llvm.minimum.* intrinsic)

in :
https://llvm.org/docs/LangRef.html

I dont understand why this is not equivalent. The ir  in the tests is atomicrmw fmin, should it not be atomicrmw and be llvm.min? Just in case:
 __c11_atomic_fetch_min(sll, 42.1, memory_order_acquire);
it generating an  atomicrmw fmin ir intrinsic, is this correct?

CarolineConcatto replied:
Ok, I saw now that using llvm.maximum and llvm.minimum we can still lower to atomic instructions. I did know that before.

Amichaxx committed:
- Formatting
- Amended logic to allow Clang to lower to expected intrinsics
- Added Sema tests to atomic-ops-fp-minmax.c

CarolineConcatto commented:
@Amichaxx I think you need to update your tests. It is failing:
TEST 'Clang :: Sema/atomic-ops.c' FAILED *
You should re-generate the CHECK lines.

Amichaxx committed:
Accidental newline

Amichaxx commented on clang/lib/CodeGen/CGAtomic.cpp:
```diff
@@ -507,20 +507,35 @@ static llvm::Value *EmitPostAtomicMinMax(CGBuilderTy &Builder,
+  llvm::Type *ValTy = OldVal->getType();
+
   llvm::CmpInst::Predicate Pred;
+  bool IsFP = ValTy->isFloatingPointTy();
```
Thank you. Amended now.

Amichaxx commented:
@CarolineConcatto Thanks for pointing that out, but I am a bit apprehensive to add Tim as he hasn't contributed to the project in nearly 2 years. Otherwise, I have added Sema tests in a separate file.

Amichaxx commented:
@efriedma-quic Hi, I've since amended as per your review. Does it look okay? Thanks.

efriedma-quic reviewed (approved):
LGTM with one minor comment.

Amichaxx committed:
Co-authored-by: Eli Friedman <efriedma@quicinc.com>

Use llvm::FMFSource() in CreateBinaryIntrinsic call

Amichaxx committed:
Clang format

CarolineConcatto reviewed (approved):
Thank you Amina, can you merge the PR?

Amichaxx committed:
Merge branch 'main' into atomics-crash-2024
